File: Mobile/lib/utils/cancellation_policy.dart
// Customer cancellation rules — mirrors shared/cancellationPolicy.ts and backend util.

class CancellationPolicyConfig {
  const CancellationPolicyConfig({required this.allowed, required this.hoursBefore});

  final bool allowed;
  final int hoursBefore;

  factory CancellationPolicyConfig.fromBusiness(Map<String, dynamic>? business) {
    if (business == null) {
      return const CancellationPolicyConfig(allowed: true, hoursBefore: 24);
    }
    final allowed = business['cancellationAllowed'] != false;
    final raw = business['cancellationHoursBefore'];
    final hours = raw is num ? raw.toInt() : int.tryParse('$raw') ?? 24;
    return CancellationPolicyConfig(allowed: allowed, hoursBefore: hours < 0 ? 24 : hours);
  }
}

String formatCancellationPolicyMessage(CancellationPolicyConfig policy, {bool isEn = true}) {
  if (!policy.allowed) {
    return isEn
        ? 'This business does not allow cancellations.'
        : 'Este negocio no permite cancelaciones.';
  }
  if (policy.hoursBefore <= 0) {
    return isEn
        ? 'You may cancel anytime before your appointment.'
        : 'Puedes cancelar en cualquier momento antes de tu cita.';
  }
  return isEn
      ? 'Cancellations must be done ${policy.hoursBefore}h before.'
      : 'Las cancelaciones deben hacerse ${policy.hoursBefore}h antes.';
}

String policyMessageForBooking({
  required String status,
  required DateTime? appointmentAt,
  String? transactionId,
  CancellationPolicyConfig? policy,
  bool isEn = true,
}) {
  final s = status.toLowerCase();
  if (s == 'pending' && (transactionId == null || transactionId.isEmpty)) {
    return isEn
        ? 'You may cancel before the venue accepts your booking.'
        : 'Puedes cancelar antes de que el negocio acepte tu reserva.';
  }
  return formatCancellationPolicyMessage(policy ?? const CancellationPolicyConfig(allowed: true, hoursBefore: 24), isEn: isEn);
}

bool canCustomerCancelBooking({
  required String rawStatus,
  required DateTime? appointmentAt,
  String? transactionId,
  CancellationPolicyConfig? policy,
  DateTime? now,
}) {
  final status = rawStatus.trim();
  if (status == 'Completed' || status == 'Cancelled' || status == 'Rejected') {
    return false;
  }
  if (status == 'Pending' && (transactionId == null || transactionId.isEmpty)) {
    return true;
  }
  final p = policy ?? const CancellationPolicyConfig(allowed: true, hoursBefore: 24);
  if (!p.allowed) return false;
  if (appointmentAt == null || p.hoursBefore <= 0) return true;
  final clock = now ?? DateTime.now();
  final deadline = appointmentAt.subtract(Duration(hours: p.hoursBefore));
  return !clock.isAfter(deadline);
}

File: Web/src/lib/cancellationPolicy.ts
export type CancellationPolicyConfig = {
  allowed: boolean;
  hoursBefore: number;
};

export function normalizeCancellationPolicy(
  business?: {
    cancellationAllowed?: boolean | null;
    cancellationHoursBefore?: number | null;
  } | null,
): CancellationPolicyConfig {
  const allowed = business?.cancellationAllowed !== false;
  const raw = business?.cancellationHoursBefore;
  const hoursBefore =
    typeof raw === "number" && Number.isFinite(raw) && raw >= 0 ? Math.floor(raw) : 1;
  return { allowed, hoursBefore };
}

export function formatCancellationPolicyMessage(
  policy: CancellationPolicyConfig,
  lang: "en" | "es" = "en",
): string {
  if (!policy.allowed) {
    return lang === "es"
      ? "Este negocio no permite cancelaciones."
      : "This business does not allow cancellations.";
  }
  if (policy.hoursBefore <= 0) {
    return lang === "es"
      ? "Puedes cancelar o reagendar tu reserva sin costo hasta 60 minutos antes de la hora programada."
      : "You may cancel or reschedule at no cost up to 60 minutes before your appointment.";
  }
  const minutes = policy.hoursBefore * 60;
  if (minutes <= 60) {
    return lang === "es"
      ? "Puedes cancelar o reagendar tu reserva sin costo hasta 60 minutos antes de la hora programada."
      : "You may cancel or reschedule at no cost up to 60 minutes before your appointment.";
  }
  return lang === "es"
    ? `Las cancelaciones deben hacerse ${policy.hoursBefore}h antes de tu cita.`
    : `Cancellations must be done ${policy.hoursBefore}h before your appointment.`;
}

export function canCustomerCancelBooking(input: {
  status: string;
  appointmentAt: string | Date;
  transactionId?: string | null;
  business?: {
    cancellationAllowed?: boolean | null;
    cancellationHoursBefore?: number | null;
  } | null;
  now?: Date;
}): { allowed: boolean; message: string } {
  const status = String(input.status || "");
  const now = input.now ?? new Date();
  if (status === "Completed" || status === "Cancelled" || status === "Rejected") {
    return { allowed: false, message: "This booking can no longer be cancelled." };
  }
  const isPendingUnpaid = status === "Pending" && !input.transactionId;
  if (isPendingUnpaid) {
    return {
      allowed: true,
      message: "You may cancel before the venue accepts your booking.",
    };
  }
  const policy = normalizeCancellationPolicy(input.business);
  if (!policy.allowed) {
    return { allowed: false, message: "This business does not allow cancellations." };
  }
  const appointmentAt =
    typeof input.appointmentAt === "string"
      ? new Date(input.appointmentAt)
      : input.appointmentAt;
  if (!Number.isNaN(appointmentAt.getTime()) && policy.hoursBefore > 0) {
    const deadline = new Date(
      appointmentAt.getTime() - policy.hoursBefore * 60 * 60 * 1000,
    );
    if (now.getTime() > deadline.getTime()) {
      return {
        allowed: false,
        message: `Cancellations must be made at least ${policy.hoursBefore} hours before the appointment.`,
      };
    }
  }
  return { allowed: true, message: "Cancellation is allowed." };
}

export function policyMessageForBooking(
  input: {
    status: string;
    appointmentAt: string | Date;
    transactionId?: string | null;
    business?: {
      cancellationAllowed?: boolean | null;
      cancellationHoursBefore?: number | null;
    } | null;
  },
  lang: "en" | "es",
): string {
  const status = String(input.status || "").toLowerCase();
  if (status === "pending" && !input.transactionId) {
    return lang === "es"
      ? "Puedes cancelar o reagendar tu reserva sin costo hasta 60 minutos antes de la hora programada."
      : "You may cancel or reschedule at no cost up to 60 minutes before your appointment.";
  }
  return formatCancellationPolicyMessage(normalizeCancellationPolicy(input.business), lang);
}
